client.go

```diff
@@ -27,11 +28,15 @@ type Conn struct {
 
 // A Client defines parameters for a DNS client.
 type Client struct {
-	Net            string            // if "tcp" or "tcp-tls" (DNS over TLS) a TCP query will be initiated, otherwise an UDP one (default is "" for UDP)
-	UDPSize        uint16            // minimum receive buffer for UDP messages
-	TLSConfig      *tls.Config       // TLS connection configuration
-	Timeout        time.Duration     // a cumulative timeout for dial, write and read, defaults to 0 (disabled) - overrides DialTimeout, ReadTimeout and WriteTimeout when non-zero
-	DialTimeout    time.Duration     // net.DialTimeout, defaults to 2 seconds - overridden by Timeout when that value is non-zero
+	Net       string      // if "tcp" or "tcp-tls" (DNS over TLS) a TCP query will be initiated, otherwise an UDP one (default is "" for UDP)
+	UDPSize   uint16      // minimum receive buffer for UDP messages
+	TLSConfig *tls.Config // TLS connection configuration
+	Dialer    *net.Dialer // a net.Dialer used to set local address, timeouts and more
```

Comment:
I don't see the change? This should be `Dialer net.Dialer`?

Reply:
there is intentionally no change, I explained above why I used a pointer. Reporting here for ease:

```
it's a pointer so it can be nil, which means "I don't want to set any Dialer, use your defaults".

In Dial I do this:

    var d net.Dialer 
    if c.Dialer == nil {
        d = net.Dialer{}
    } else { 
        d = net.Dialer(*c.Dialer) 
    } 
which eventually uses another dialer instance - the one from Client, if any, or a default one if not
```

Reply:
In other words I want `Dialer` to be `nil` by default, in which case the defaults are used. If set, the defaults will be overridden

Reply:
@insomniacslk Using a pointer isn't needed to do that. If it's `nil`, you're just explicitly using the zero value anyway. The following would achieve the exact same operation and is less code; I don't see why a pointer is preferable:

```
type Client struct {
	...
	Dialer net.Dialer
	...
}

func (c *Client) Dial(address string) (conn *Conn, err error) {
	d := c.Dialier // this creates a copy of Dialer
	...
}
```